services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/utils/MongoDbDocumentBuilder.java

```diff
@@ -129,10 +148,57 @@ public MongoDbDocumentBuilder withType(final String type) {
      * @param authId The auth id.
      * @return a reference to this for fluent use.
      */
-    public MongoDbDocumentBuilder withAuthId(final String authId) {
+    private MongoDbDocumentBuilder withAuthId(final String authId) {
         return withCredentialsPredicate(RegistryManagementConstants.FIELD_AUTH_ID, authId);
     }
 
+    /**
+     * Sets the json object with the given auth id.
+     * <p>
+     * A MongoDB query is framed to match the given auth id to
+     * <ul>
+     * <li>the field <em>generated-auth-id</em>.</li>
+     * <li>If no match found and <em>generated-auth-id</em> is {@code null}, match with the field <em>auth-id</em>.
+     * </li>
+     * </ul>
+     * <p>
+     * Example query with auth id as "Device1-Hono-Eclipse":
+     * <pre>
+     * {
+     *     "credentials": {
+     *         "$elemMatch": {
+     *             "$or": [
+     *                 {
+     *                     "generated-auth-id": "Device1-Hono-Eclipse"
+     *                 },
+     *                 {
+     *                     "generated-auth-id": null,
+     *                     "auth-id": "Device1-Hono-Eclipse"
+     *                 }
+     *             ]
+     *         }
+     *     }
+     * }
+     * </pre>
+     *
+     * @param authId The auth id.
+     * @return a reference to this for fluent use.
+     */
+    private MongoDbDocumentBuilder withGeneratedAuthId(final String authId) {
+        final var credentialsArraySpec = document.getJsonObject(CredentialsDto.FIELD_CREDENTIALS, new JsonObject());
+        final var elementMatchSpec = credentialsArraySpec.getJsonObject(MONGODB_OPERATOR_ELEM_MATCH, new JsonObject());
+
+        elementMatchSpec.put(MONGODB_OPERATOR_OR, new JsonArray(List.of(
```

Comment:
I think a List is needed there as there are two JsonObjects to be added to the JsonArray.  If we like to avoid using List, then it can be rewritten as below. One way of populating JsonArray is to pass a list to the constructor and another is to invoke add methods.

```
elementMatchSpec.put(MONGODB_OPERATOR_OR, new JsonArray().add(new JsonObject()).add(new JsonObject())
```